creed/tiling_utils_test.py

```diff
@@ -0,0 +1,745 @@
+        "Sample prompt: ",
+        "Sample error message",
+        lambda x: x == 1,
+        dtype=int
+    )
+
+    # assert sample_tiling_param was set to 1
+    assert sample_tiling_param == 1
+
+    # test 2: str inputs
+    # test an invalid non-str response, then an invalid str response, then a valid response
+    user_inputs_str = iter([1, 'N', 'Y'])
+
+    # make sure the function receives the incorrect input first then the correct input
+    monkeypatch.setattr('builtins.input', lambda _: next(user_inputs_str))
+
+    # simulate the input sequence for str
+    sample_tiling_param = tiling_utils.read_tiling_param(
+        "Sample prompt: ",
+        "Sample error message",
+        lambda x: x == 'Y',
+        dtype=str
+    )
+
+    # assert sample_tiling_param was set to 'Y'
+    assert sample_tiling_param == 'Y'
+
+
+def test_read_tiled_region_inputs(monkeypatch):
+    # define a sample fovs list
+    sample_fovs_list = test_utils.generate_sample_fovs_list(
+        fov_coords=[(0, 150), (100, 300)], fov_names=["TheFirstFOV", "TheSecondFOV"]
+    )
+
+    # define sample region_params to read data into
+    sample_region_params = {rpf: [] for rpf in settings.REGION_PARAM_FIELDS}
+
+    # intentionally place some lowercase letters as this function should support those
+    user_inputs = iter([2, 4, 1, 2, 'N', 4, 8, 2, 4, 'y'])
+
+    # override the default functionality of the input function
+    monkeypatch.setattr('builtins.input', lambda _: next(user_inputs))
+
+    # use the dummy user data to read values into the params lists
+    tiling_utils.read_tiled_region_inputs(
+        sample_fovs_list, sample_region_params
+    )
+
+    assert (sample_region_params[i]['region_start_x'] == 100 * i
+            for i in range(len(sample_region_params)))
+    assert (sample_region_params[i]['region_start_y'] == 150 * (i + 1)
+            for i in range(len(sample_region_params)))
+    assert (sample_region_params[i]['fov_num_x'] == 2 * (i + 1)
+            for i in range(len(sample_region_params)))
+    assert (sample_region_params[i]['fov_num_y'] == 4 * (i + 1)
+            for i in range(len(sample_region_params)))
+    assert (sample_region_params[i]['x_fov_size'] == 1 * (i + 1)
+            for i in range(len(sample_region_params)))
+    assert (sample_region_params[i]['y_fov_size'] == 2 * (i + 1)
+            for i in range(len(sample_region_params)))
+
+    assert sample_region_params['region_rand'] == ['N', 'Y']
+
+
+def test_generate_region_info():
+    sample_region_inputs = {
+        'region_start_x': [100, 200],
+        'region_start_y': [200, 400],
+        'fov_num_x': [3, 6],
+        'fov_num_y': [6, 12],
+        'x_fov_size': [5, 10],
+        'y_fov_size': [10, 20],
+        'region_rand': ['N', 'Y']
+    }
+
+    # generate the region params
+    sample_region_params = tiling_utils.generate_region_info(sample_region_inputs)
+
+    # assert region_start_x set correctly
```

Comment:
These assert statements can be condensed into a for loop over the tuple of parameter (i.e. region_start_x), base_value (i.e. 100)

Reply:
Were you thinking of something more or less like:

```
# define the starting values to increment from for each param
base_vals = {
    'region_start_x': 100,
    'region_start_y': 200,
    'fov_num_x': 3,
    'fov_num_y': 6,
    'x_fov_size': 5,
    'y_fov_size': 10
}

# the keys are going to be the same for all regions of sample_region_params
for i in list(zip(sample_region_params[0].items()))
    for param, val in list(sample_region_params[i].items()):
        assert val == base_vals[param] * (i + 1)
```

Reply:
Yup, and you could use base_vals to generate sample_region_inputs as well